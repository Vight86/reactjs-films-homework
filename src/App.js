import { hot } from 'react-hot-loader/root';
import React, { useState } from 'react';
import { useDispatch, useSelector, shallowEqual } from 'react-redux';
import {
  Redirect, Route, useRouteMatch,
} from 'react-router-dom';

import { push } from 'connected-react-router';

import { selectQuery } from './modules/query/index';
import { addCurrentMovieTrailerKey, currentMovieTrailerKeyAdded, selectMovies } from './modules/movies/index';

import ErrorBoundary from './components/ErrorBoundary/index';
import Header from './components/Header/index';
import Logo from './components/Logo/index';
import SearchForm from './components/SearchForm/index';
import SearchResultsPageContainer from './pages/SearchResultsPage/index';
import MovieDetailsPageContainer from './pages/MovieDetailsPage/index';
import Footer from './components/Footer/index';
import Modal from './components/Modal/index';

const App = () => {
  const dispatch = useDispatch();

  const handleSearchSubmit = (query, e) => {
    e.preventDefault();
    dispatch(push(`?query=${query}`));
  };

  const { basePath, apiKey } = useSelector(selectQuery);
  const updateMovieTrailerKey = (id) => {
    if (id === null) {
      dispatch(currentMovieTrailerKeyAdded(id));
      return;
    }
    const url = `${basePath}/movie/${id}/videos?api_key=${apiKey}`;
    dispatch(addCurrentMovieTrailerKey(url));
  };

  const [isModalOpened, setIsModalOpened] = useState(false);
  const handleModalToggle = (movieId, isOpened) => {
    const paddingRight = window.innerWidth - document.documentElement.clientWidth;
    updateMovieTrailerKey(movieId);
    setIsModalOpened(!isOpened);
    document.body.style.overflow = !isOpened ? 'hidden' : '';
    document.body.style.paddingRight = !isOpened ? `${paddingRight}px` : '0px';
  };

  const { currentMovieTrailerKey } = useSelector(selectMovies, shallowEqual);

  return (
    <>
      <ErrorBoundary>
        <Header className="primary">
          <Logo className="primary">Films</Logo>
          <SearchForm handleSearchSubmit={handleSearchSubmit} />
        </Header>
      </ErrorBoundary>
      <ErrorBoundary>
        <Route path="/:category/movie/:id">
          <MovieDetailsPageContainer
            isModalOpened={isModalOpened}
            handleModalToggle={handleModalToggle}
          />
        </Route>
      </ErrorBoundary>
      <ErrorBoundary>
        <Route path="/:category">
          <SearchResultsPageContainer
            isModalOpened={isModalOpened}
            handleModalToggle={handleModalToggle}
          />
        </Route>
      </ErrorBoundary>
      <Footer />
      <ErrorBoundary>
        <Modal
          movieTrailerKey={currentMovieTrailerKey}
          isModalOpened={isModalOpened}
          handleModalToggle={handleModalToggle}
        />
      </ErrorBoundary>
      {useRouteMatch().isExact ? <Redirect to="/popular" /> : null}
    </>

  );
};

export default hot(App);
