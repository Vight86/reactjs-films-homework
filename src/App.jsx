import { hot } from 'react-hot-loader/root';
import React from 'react';
import Header from './components/Header/index';
import Logo from './components/Logo/index';
import SearchForm from './components/SearchForm/index';
import SearchResultsPage from './pages/SearchResultsPage/index';
import Footer from './components/Footer/index';

const App = () => (
  <>
    <Header className="primary">
      <Logo />
      <SearchForm />
    </Header>
    <SearchResultsPage />
    <Footer />
  </>
);

export default hot(App);
