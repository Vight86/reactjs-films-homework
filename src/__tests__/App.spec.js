import React from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import App from '../App';

let mockAppState;
jest.mock('react-redux', () => ({
  ...jest.requireActual('react-redux'),
  useDispatch: jest.fn(() => jest.fn()),
  useSelector: jest.fn().mockImplementation((callback) => callback(mockAppState)),
}));

let mockIsExact;
jest.mock('react-router-dom', () => ({
  ...jest.requireActual('react-redux'),
  useRouteMatch: jest.fn(() => ({ isExact: mockIsExact })),
  Redirect: jest.fn(),
  Route: jest.fn(),
}));

let renderer;
let result;

beforeEach(() => {
  mockAppState = {
    query: {
      basePath: 'http://testPath.com',
      apiKey: 'testKey',
    },
    movies: {
      currentMovieTrailerKey: 'test',
    },
  };
  mockIsExact = true;
  renderer = new ShallowRenderer();
  renderer.render(<App />);
  result = renderer.getRenderOutput();
});

afterEach(() => {
  mockAppState = null;
  mockIsExact = null;
  renderer = null;
  result = null;
});

describe('render App component', () => {
  it('App renders correctly', () => {
    expect(result).toMatchSnapshot();
  });

  it('renders no redirect from inner routes', () => {
    mockIsExact = false;
    renderer.render(<App />);
    result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('renders handleSearchSubmit correctly', () => {
    const searchForm = result.props.children[0].props.children.props.children[1];
    const mockQuery = 'test';
    const mockEvent = { preventDefault: jest.fn() };
    const { handleSearchSubmit } = searchForm.props;
    handleSearchSubmit(mockQuery, mockEvent);
    result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('renders handleModalToggle to open modal correctly', () => {
    const { handleModalToggle } = result.props.children[4].props.children.props;
    handleModalToggle('testId', true);
    result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });

  it('renders handleModalToggle to close modal correctly', () => {
    const { handleModalToggle } = result.props.children[4].props.children.props;
    handleModalToggle(null, false);
    result = renderer.getRenderOutput();
    expect(result).toMatchSnapshot();
  });
});
