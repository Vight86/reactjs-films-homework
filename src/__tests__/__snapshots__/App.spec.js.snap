// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders correctly 1`] = `
<React.Fragment>
  <Header
    className="primary"
  >
    <Logo
      theme="primary"
    />
    <SearchForm />
  </Header>
  <MovieDetailsPage />
  <Footer />
</React.Fragment>
`;
